documentation/Sprint4.md

```diff
@@ -0,0 +1,416 @@
+- Opens the channel menu when the channel button is clicked
+- Selects a channel when a menu item is clicked
+- Marks the current selected channel as selected in the menu
+- Handles the case when selected Server has no channels
+- Handles the case when selectedServer is null
+
+**ServerAndMembers.test**
+- Renders without crashing
+- Passes the correct props to ActiveServer
+- Does not render Members by default (visible is false)
+- Renders Members when visible state is true
+- Forwards channel selection to parent component
+- Handles null server and channel gracefully
+
+**ServerBadge.test**
+- Renders without exceptions
+- Renders server name
+- Renders default server icon
+- Renders provided server icon
+
+**ServerList.test**
+- Renders without exceptions
+- Server list renders all test servers
+- Search filters servers returned
+- Selecting a server calls onServerSelect and onChannelSelect with default channel
+- Clicking on a channel calls onChannelSelect
+- Adding a new channel works
+- Visual indication is applied to the selected channel
+
+**UserBadge.test**
+- Renders without exceptions
+- Renders server name
+- Renders default server icon
+- Renders provided server icon
+
+## Cypress Tests
+
+- Should display login screen on initial load
+- Should require both username and password to login
+- Should log in with valid credentials and show the main screen
+- Should display server list with multiple servers
+- Should select a server and display its channels
+- Should allow adding a new channel to a server
+- Should switch between channels and display correct messages
+- Should send a new message in a channel
+- Should open the user profile edit modal
+
+
+# Backend
+## **Backend Implementation**
+
+## **Authentication System**
+- **JWT Authentication**:  
+  Implemented a robust **JSON Web Token (JWT)** authentication system with secure token generation, validation, and management.
+  
+- **Challenge-Response Authentication**:  
+  Developed a secure challenge-response mechanism for user login using **public key cryptography**.
+  
+- **Session Management**:  
+  Created comprehensive session handling with **token expiration** and **refresh capabilities**.
+
+## **Encryption**
+- **End-to-End Encryption Support**:  
+  Implemented backend support for **end-to-end encrypted messaging**.
+  
+- **Key Management**:  
+  Developed secure storage and retrieval mechanisms for user **public keys** used in message encryption.
+  
+- **Message Integrity**:  
+  Added verification mechanisms to ensure **message integrity** during transmission.
+
+## **Database Enhancements**
+- **Optimized Queries**:  
+  Improved **database query performance** for faster message and channel retrieval.
+  
+- **Data Consistency**:  
+  Enhanced **data consistency checks** during create, update, and delete operations.
+  
+- **Relationship Management**:  
+  Strengthened **relationship handling** between users, groups, channels, and messages.
```

Comment:
Technically we didn't do this, I mean we maybe kinda did unintentionally through some debugging stuff so might be fine